programs/drift_vaults/src/instructions/redeem_tokens.rs

```diff
@@ -0,0 +1,204 @@
+        &mut vault,
+        total_supply_before,
+        vault_equity,
+        tokens_to_burn,
+        clock.unix_timestamp,
+    )?;
+
+    let shares_transferred = tokenized_vault_depositor.transfer_shares(
+        &mut *vault_depositor,
+        &mut vault,
+        shares_to_transfer,
+        WithdrawUnit::Shares,
+        vault_equity,
+        clock.unix_timestamp,
+    )?;
+
+    let total_shares_after = vault_depositor
+        .get_vault_shares()
+        .safe_add(tokenized_vault_depositor.get_vault_shares())?;
+
+    validate!(
+        total_shares_after.eq(&total_shares_before),
+        ErrorCode::InvalidVaultSharesDetected,
+        "Total vault depositor shares before != after"
+    )?;
+
+    validate!(
+        shares_transferred == shares_to_transfer.into(),
+        ErrorCode::InvalidVaultSharesDetected
+    )?;
+
+    let vault_name = vault.name;
+    let vault_bump = vault.bump;
+
+    drop(vault);
+    drop(vault_depositor);
+    drop(tokenized_vault_depositor);
+
+    ctx.token_transfer(tokens_to_burn)?;
+    ctx.burn(vault_name, vault_bump, tokens_to_burn)?;
+
+    msg!(
+        "Burned {} tokens from {}",
+        tokens_to_burn,
+        ctx.accounts.user_token_account.key()
+    );
+
+    ctx.accounts.mint.reload()?;
+    let total_supply_after = ctx.accounts.mint.supply;
+
+    validate!(
+        total_supply_after < total_supply_before,
+        ErrorCode::InvalidTokenization,
+        "Total supply after > total supply before"
+    )?;
+
+    let supply_delta = total_supply_before.safe_sub(total_supply_after)?;
+    validate!(
+        supply_delta.eq(&tokens_to_burn),
+        ErrorCode::InvalidTokenization,
+        "Tokens burned ({}) != supply delta ({})",
+        tokens_to_burn,
+        supply_delta
+    )?;
+
+    Ok(())
+}
+
+#[derive(Accounts)]
+pub struct RedeemShares<'info> {
+    #[account(mut)]
+    pub vault: AccountLoader<'info, Vault>,
+    #[account(
+        mut,
+        seeds = [b"vault_depositor", vault.key().as_ref(), authority.key().as_ref()],
+        bump,
+        constraint = is_authority_for_vault_depositor(&vault_depositor, &authority)?,
+    )]
+    pub vault_depositor: AccountLoader<'info, VaultDepositor>,
+    pub authority: Signer<'info>,
+    #[account(
+		mut,
+		constraint = is_tokenized_depositor_for_vault(&tokenized_vault_depositor, &vault)?,
+	)]
+    pub tokenized_vault_depositor: AccountLoader<'info, TokenizedVaultDepositor>,
+    #[account(
+        mut,
+        seeds = [b"mint", vault.key().as_ref()],
+        bump,
+        mint::authority = vault.key(),
+		constraint = is_mint_for_tokenized_depositor(&mint.key(), &tokenized_vault_depositor)?,
+    )]
+    pub mint: Account<'info, Mint>,
+    #[account(
+        mut,
+        token::authority = authority,
+        token::mint = tokenized_vault_depositor.load()?.mint
+    )]
+    pub user_token_account: Account<'info, TokenAccount>,
+    #[account(
+        mut,
```

Comment:
you might wanna make sure the vault is the program initialized

im not sure if there is a way to take advantage of it, but somebody could make a diff token account with the vault as authority